generative/maisi/scripts/train_controlnet.py

```diff
@@ -0,0 +1,427 @@
+        list_train = partition_dataset(
+            data=list_train,
+            shuffle=True,
+            num_partitions=world_size,
+            even_divisible=True,
+        )[rank]
+    train_ds = CacheDataset(data=list_train, transform=train_transforms, cache_rate=cache_rate, num_workers=8)
+    train_loader = DataLoader(train_ds, batch_size=batch_size, shuffle=True, num_workers=8, pin_memory=True)
+    if ddp_bool:
+        list_valid = partition_dataset(
+            data=list_valid,
+            shuffle=True,
+            num_partitions=world_size,
+            even_divisible=False,
+        )[rank]
+    val_ds = CacheDataset(
+        data=list_valid,
+        transform=val_transforms,
+        cache_rate=cache_rate,
+        num_workers=8,
+    )
+    val_loader = DataLoader(val_ds, batch_size=batch_size, shuffle=False, num_workers=2, pin_memory=False)
+    return train_loader, val_loader
+
+
+def main():
+    parser = argparse.ArgumentParser(description="maisi.controlnet.training")
+    parser.add_argument(
+        "-e",
+        "--environment-file",
+        default="./config/environment_maisi_controlnet_train.json",
+        help="environment json file that stores environment path",
+    )
+    parser.add_argument(
+        "-c",
+        "--config-file",
+        default="./config/config_maisi_controlnet_train.json",
+        help="config json file that stores hyper-parameters",
+    )
+    parser.add_argument("-g", "--gpus", default=1, type=int, help="number of gpus per node")
+    args = parser.parse_args()
+
+    # Step 0: configuration
+    logger = logging.getLogger("maisi.controlnet.training")
+    # whether to use distributed data parallel
+    ddp_bool = args.gpus > 1
+    if ddp_bool:
+        rank = int(os.environ["LOCAL_RANK"])
+        world_size = int(os.environ["WORLD_SIZE"])
+        device = setup_ddp(rank, world_size)
+        logger.addFilter(RankFilter())
+    else:
+        rank = 0
+        world_size = 1
+        device = torch.device(f"cuda:{rank}")
+
+    torch.cuda.set_device(device)
+    logger.info(f"Number of GPUs: {torch.cuda.device_count()}")
+    logger.info(f"World_size: {world_size}")
+
+    env_dict = json.load(open(args.environment_file, "r"))
+    config_dict = json.load(open(args.config_file, "r"))
+
+    for k, v in env_dict.items():
+        setattr(args, k, v)
+    for k, v in config_dict.items():
+        setattr(args, k, v)
+
+    # initialize tensorboard writer
+    if rank == 0:
+        tensorboard_path = os.path.join(args.tfevent_path, args.exp_name)
+        Path(tensorboard_path).mkdir(parents=True, exist_ok=True)
+        tensorboard_writer = SummaryWriter(tensorboard_path)
+
+    # Step 1: set data loader
+    train_loader, _ = prepare_maisi_controlnet_json_dataloader(
+        json_data_list=args.json_data_list,
+        data_base_dir=args.data_base_dir,
+        rank=rank,
+        world_size=world_size,
+        batch_size=args.controlnet_train["batch_size"],
```

Comment:
Please also make sure the args are parsed correctly. I didn't find it directly in the argument list.

Reply:
Those arguments are stored in configs/config_maisi_controlnet_train.json. https://github.com/Project-MONAI/tutorials/blob/40526e05fc5e2ec6d370c51de7f36f5426704edb/generative/maisi/configs/config_maisi_controlnet_train.json#L70